genie-agent/src/test/groovy/com/netflix/genie/agent/execution/services/impl/grpc/GRpcJobTerminationListenerServiceImplSpec.groovy

```diff
@@ -0,0 +1,133 @@
+/*
+ *
+ *  Copyright 2018 Netflix, Inc.
+ *
+ *     Licensed under the Apache License, Version 2.0 (the "License");
+ *     you may not use this file except in compliance with the License.
+ *     You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *     Unless required by applicable law or agreed to in writing, software
+ *     distributed under the License is distributed on an "AS IS" BASIS,
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ *
+ */
+
+package com.netflix.genie.agent.execution.services.impl.grpc
+
+import com.netflix.genie.agent.execution.ExecutionContext
+import com.netflix.genie.proto.*
+import com.netflix.genie.test.categories.UnitTest
+import io.grpc.Status
+import io.grpc.StatusRuntimeException
+import io.grpc.stub.StreamObserver
+import io.grpc.testing.GrpcServerRule
+import org.junit.Rule
+import org.junit.experimental.categories.Category
+import spock.lang.Specification
+
+@Category(UnitTest.class)
+class GRpcJobTerminationListenerServiceImplSpec extends Specification {
+
+    @Rule
+    GrpcServerRule grpcServerRule = new GrpcServerRule().directExecutor()
+    String jobId
+    GRpcJobTerminationListenerServiceImpl service
+    JobTerminationServiceGrpc.JobTerminationServiceFutureStub client;
+    ExecutionContext executionContext;
+    //A callback provided by the tests to simulate failure or success during registration
+    //by the TestJobTerminationService
+    Runnable simulationCallback;
+
+    void setup() {
+        this.grpcServerRule.getServiceRegistry().addService(new TestJobTerminationService())
+        this.jobId = UUID.randomUUID().toString()
+        executionContext = Mock()
+        executionContext.getClaimedJobId() >> jobId
+        this.client = JobTerminationServiceGrpc.newFutureStub(grpcServerRule.getChannel())
+        this.service = new GRpcJobTerminationListenerServiceImpl(client, executionContext)
+    }
+
+    def "Successfully handle kill notification from the service"() {
+        Runnable onSuccessCallback = Mock();
+        simulationCallback = new Runnable() {
+            @Override
+            void run() {}
+        }
+
+        when:
+        service.registerForRemoteKillNotification(onSuccessCallback)
+
+        then: "onSuccess callback gets invoked"
```

Comment:
If you're going to put a label on the there you should probably put a label on the `when` as well. It should tell a story ideally.  `when X happens then Y happens`

Reply:
👍